sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2178,25 +2179,87 @@ std::pair<RT::PiKernel, std::mutex *> ProgramManager::getOrCreateKernel(
+    return (PropIt == PropRange.end())
+               ? std::nullopt
+               : std::optional<
+                     RTDeviceBinaryImage::PropertyRange::ConstIterator>{PropIt};
+  };
+
+  auto AspectsPropIt = getPropIt("aspects");
+  auto ReqdWGSizePropIt = getPropIt("reqd_work_group_size");
+
+  if (!AspectsPropIt && !ReqdWGSizePropIt)
     return true;
-  ByteArray Aspects = DeviceBinaryProperty(*PropIt).asByteArray();
-  // Drop 8 bytes describing the size of the byte array.
-  Aspects.dropBytes(8);
-  while (!Aspects.empty()) {
-    aspect Aspect = Aspects.consume<aspect>();
-    if (!Dev.has(Aspect))
+
+  // Checking if device supports defined aspects
+  if (AspectsPropIt) {
+    ByteArray Aspects =
+        DeviceBinaryProperty(*(AspectsPropIt.value())).asByteArray();
+    // Drop 8 bytes describing the size of the byte array.
+    Aspects.dropBytes(8);
+    while (!Aspects.empty()) {
+      aspect Aspect = Aspects.consume<aspect>();
+      if (!Dev.has(Aspect))
+        return false;
+    }
+  }
+
+  // Checking if device supports defined required work group size
+  if (ReqdWGSizePropIt) {
+    ByteArray ReqdWGSize =
+        DeviceBinaryProperty(*(ReqdWGSizePropIt.value())).asByteArray();
+    // Drop 8 bytes describing the size of the byte array.
+    ReqdWGSize.dropBytes(8);
+    int ReqdWGSizeAllDimsTotal = 1;
+    std::vector<int> ReqdWGSizeVec;
+    int Dims = 0;
+    while (!ReqdWGSize.empty()) {
+      int SingleDimSize = ReqdWGSize.consume<int>();
+      ReqdWGSizeAllDimsTotal *= SingleDimSize;
+      ReqdWGSizeVec.push_back(SingleDimSize);
+      Dims++;
+    }
+    if (static_cast<long unsigned int>(ReqdWGSizeAllDimsTotal) >
+        Dev.get_info<info::device::max_work_group_size>())
       return false;
+    // Creating std::variant to call max_work_item_sizes one time to avoid
+    // performance drop
+    std::variant<id<1>, id<2>, id<3>> MaxWorkItemSizesVariant;
+    if (Dims == 1)
+      MaxWorkItemSizesVariant =
+          Dev.get_info<info::device::max_work_item_sizes<1>>();
+    else if (Dims == 2)
+      MaxWorkItemSizesVariant =
+          Dev.get_info<info::device::max_work_item_sizes<2>>();
+    else // (Dims == 3)
+      MaxWorkItemSizesVariant =
+          Dev.get_info<info::device::max_work_item_sizes<3>>();
+    std::vector<int> MaxWorkItemSizes;
+    // Extracting value from std::variant to avoid dealing with type-safety
+    // issues after that
+    for (int i = 0; i < Dims; i++) {
+      if (Dims == 1)
+        MaxWorkItemSizes.push_back(std::get<id<1>>(MaxWorkItemSizesVariant)[i]);
+      else if (Dims == 2)
+        MaxWorkItemSizes.push_back(std::get<id<2>>(MaxWorkItemSizesVariant)[i]);
+      else // (Dims == 3)
+        MaxWorkItemSizes.push_back(std::get<id<3>>(MaxWorkItemSizesVariant)[i]);
+    }
+    for (int i = 0; i < Dims; i++) {
+      // ReqdWGSizeVec is in reverse order compared to MaxWorkItemSizes
+      if (ReqdWGSizeVec[i] > MaxWorkItemSizes[Dims - i - 1])
+        return false;
```

Comment:
Nit; would it make sense to move this into the other loop? As such we could avoid the need for the vector by just storing the value of each iteration into a temporary int and use that for the comparison.

Reply:
Applied in https://github.com/intel/llvm/pull/8056/commits/8147ca4c8b118f1be9ca62e4f28d5fc70a485832